### ISSUE TITLE
FEE over/undershoot

### ISSUE THREAD

**Issue by carstenpaproth** [IvS-KULeuven/PlatoSim3#376]
The F-FEE people noticed some over- and undershoot in their measurements. As a similar effect was visible in Kepler (?), it would be good to have a model for this effect implemented in PLATOSim, so we could check, for example, the impact on the FGS. Can someone of the PLATOSim-Team implement the over/undershoot effect in the near future?
 
Short description of the effect:
The effect is linear in the signal level change during the readout.
It should be only visible in readout direction.
It can be positive (overshoot) or negative (undershoot) depending on the sign of the signal level change.
It seems to be symmetric.
There is an exponential decay of this effect for about 4 pixels.
It is a rather small effect, for a signal level of 65000 DN, they observed about 200 DN over- and undershoot.
I tried to illustrate an exaggerated effect.

[plato_overshoot.pdf](https://github.com/IvS-KULeuven/PlatoSim3/files/3497017/plato_overshoot.pdf)

**Comment by SaraRegibo** [comment]
So for a signal change `ΔS` from one pixel `x-1` in the readout register to the next one, `x`, the over-/undershoot can be characterised as:

     [(200 / 65000) * ΔS] * exp(-λ x),

where `λ` is the rate of the decay and *x* is the pixel number in the readout register.  This should only affect pixels `x`, `x + 1`, `x + 2`, and `x + 3`.

So the required configuration parameters for PlatoSim are:
- 200 / 65000 -> do you have a better name than `N0`?
- λ -> `Rate`
- 4 pixels -> `Range`

I do have a couple of questions:
- You mention this was noticed by the F-FEE people, but I guess this effect also applies to the **N-FEE**, right?
- Do you know whether this effect has been fully characterised?  What is the **rate** of the decay?
- What happens with the **1st pixel** in the readout register?
- It's the difference in signal level that I have to use in the 1st formula (and not the signal level itself), right?
- In case you don't have the answers, whom should I ask about this?

**Comment by carstenpaproth** [comment]
I think it should be Delta x in your formula:

[(200 / 65000) * ΔS] * exp(-λ Δx)

Currently, we don’t expect this effect for the N-FEE.
The characterization of this effect is still ongoing.
I will extract the rate of the decay from brand new measurements and let you know.


Von: Sara Regibo [mailto:notifications@github.com]
Gesendet: Montag, 26. August 2019 14:46
An: IvS-KULeuven/PlatoSim3
Cc: Paproth, Carsten; Author
Betreff: Re: [IvS-KULeuven/PlatoSim3] FEE over/undershoot (#376)


So for a signal change ΔS from one pixel x-1 in the readout register to the next one, x, the over-/undershoot can be characterised as:

 [(200 / 65000) * ΔS] * exp(-λ x),



where λ is the rate of the decay and x is the pixel number in the readout register. This should only affect pixels x, x + 1, x + 2, and x + 3.

So the required configuration parameters for PlatoSim are:

  *   200 / 65000 -> do you have a better name than N0?
  *   λ -> Rate
  *   4 pixels -> Range

I do have a couple of questions:

  *   You mention this was noticed by the F-FEE people, but I guess this effect also applies to the N-FEE, right?
  *   Do you know whether this effect has been fully characterised? What is the rate of the decay?
  *   In case you don't have the answers, whom should I ask about this?

**Comment by carstenpaproth** [comment]
I analyzed the new F-FEE measurements and fitted your model into the data. To get a good fit, every readout pixel is affected by this decay of the signal change from the last 5 pixels. So, the range is 6 pixels. Furthermore, I had to use a slower decay. 

Sum_i [a * ΔS_i] * exp(-λ i^b)

The model parameters:
a = 0.003867
λ = 1.277
b = 0.755
for i = 0, 1, 2, 3, 4 and ΔS_i the signal changes between the last 5 neighboring pixel pairs,  ΔS_0 is the signal change from the previous pixel to the current pixel,  ΔS_1 is the change from the pixel before the previous pixel to the previous pixel and so on.

**Comment by SaraRegibo** [comment]
Thanks for the input, Carsten!  Can you explain to me what the physical reason is for this effect and why you only expect it for the F-FEE? Just out of curiosity...

**Comment by carstenpaproth** [comment]
I couldn't get a plausible physical explanation for this effect yet. It could be some weird behavior inside the used signal processor (VASP). Or it could be just a systematic error in the F-FEE test setup and not a real problem. Currently, there is no reason to expect a similar effect for the N-FEE.

**Comment by SaraRegibo** [comment]
Proposed parameter names (please, complete and/or correct if anyone has better ideas):
* `a` -> ?
* `b` -> ?
* `λ` -> `Rate`
* `Range`: indicating the next 5 pixels are affected
* `IncludeOverAndUnderShoot`: boolean indicating whether or not the effect should be included -> should be cross-checked with the value of the `Telescope/GroupID` parameter, which should be set to `Fast`

**Comment by SaraRegibo** [comment]
Food for thought: at which point (in the chain of effects being applied) does the over-/undershoot have to be applied?

**Comment by carstenpaproth** [comment]
a -> Strength
b -> Decay Speed
λ -> Decay Rate
Range 5 pixels

**Comment by carstenpaproth** [comment]
The effect should be applied after all signal relevant effects were applied, just before the readout noise is added.

**Comment by SaraRegibo** [comment]
- [x] Introduce new configuration parameters in the `FEE` block of all configuration files
- [x] Update `Simulation::writeInputParametersToHDF5`
- [x] Update documentation
- [ ] Update TN0006
- [ ] Update TN007
- [x] configuration (in `FrontEndElectronics`)
- [x] Implement the `applyOverAndUnderShoot()` method (in `Detector`)
- [x] Plug in the `applyOverAndUnderShoot()` method at the appropriate place (in `Detector`)

**Comment by SaraRegibo** [comment]
Whilst I am implementing, @carstenpaproth, is there document describing the origin of the proposed parameter values, to which I can refer in TN0006?

**Comment by SaraRegibo** [comment]
Other things to pay attention to:
* Also the serial pre-scan (the so-called bias map) and the parallel over-scan (the so-called smearing map) are affected...
* To account for this effect correctly, we would also have to know the value of the pixels between the bias map and the pixel map (if any).  Proposal is to assume the sky background.
* Who invented this concept of two detector halves?!

**Comment by carstenpaproth** [comment]
There is no document containing the description yet. The parameters were derived from an F-FEE test setup measurement with a synthetical Gaussian signal peak. You can refer to this measurement which was released on 26.08.2019 by the F-FEE team. If the effect turns out to be relevant for the FGS, we need to refine the parameters with proper F-FEE measurements.

**Comment by SaraRegibo** [comment]
So ideally, if we would apply the effect to the serial pre-scan as well (and propagate to the sub-field or the parallel over-scan), the content of the readout register is (speaking for the left CCD half here):
- the extra pixels at the start of the readout register, that make up the serial pre-scan;
- potentially a number of pixels that are not simulated (i.e. are not in the sub-field) -> use the sky background here (?);
- and finally the simulated pixels.

Now I realise that we introduced the `NumBiasPrescanRows` parameter to configure the number of rows in the simulated serial pre-scan (for performance reasons), but this gets us into trouble now, obviously, if the number of rows in the sub-field and in the serial pre-scan are not the same (a row in the sub-field might not have a corresponding row in the serial over-scan; so what would we put in the readout register then?).

Maybe I am overcomplicating things here.  Since it's a very small effect, it may be overkill wanting to apply it to the serial pre-scan, too.  Or should we enforce the number of rows to match (not really in favour of this)?  Maybe there are other potential solutions?

And an additional question: should we also apply it to the parallel over-scan?

**Comment by SaraRegibo** [comment]
From a discussion we had here this morning:
- Since the effect is so small, we choose not to apply it to the serial pre-scan (i.e. the bias map).
- To make sure it's (more or less) correct for the first 5 columns, we will use 5 extra pixels, with the sky background as value.
- Should we also apply it to the parallel over-scan (i.e. the smearing map)? -> might not even be necessary for the envisaged purpose.

**Comment by carstenpaproth** [comment]
I think it is fine to not apply this effect to the serial pre-scan and parallel over-scan. Instead of using the sky background for not simulated pixels, you could simply use 0 as the delta signal of not simulated pixels.

**Comment by SaraRegibo** [comment]
The order of the effects added in the `readOut(float exposureTime)` method is hence the following:
- cosmics
- open-shutter smearing
- CTI
- photon noise
- **(F-)FEE over-/undershoot**
- readout noise
- full-well saturation (aka blooming)
- quantisation
     - gain
     - electronic offset
     - flooring ADU values
     - digital saturation

**Comment by SaraRegibo** [comment]
Implemented in Pull Request #404.  For @denisGrBach and @carstenpaproth to experiment with!

**Comment by carstenpaproth** [comment]
@SaraRegibo 

Thanks for implementing the effect. We made a short test with your implementation and it produces an over/undershoot effect. Unfortunately, your implementation doesn't behave as expected:
1. For one CCD half the effect is 0
`difference = readoutRegister.tail(lengthReadoutRegister - deltaX) - readoutRegister.tail(lengthReadoutRegister - deltaX);`
2. In exp(-λ i^b), the i shall start at 0, but in your code the deltaX starts with 1
`exp(-frontEndElectronics->getOverAndUnderShootDecayRate() * pow(deltaX, frontEndElectronics->getOverAndUnderShootDecaySpeed()))`
3. The wrong ΔS_i are applied for a pixel. For example, the signal in pixel 5 shall be influenced by the signal change from pixel 4 to pixel 5, the signal change from pixel 3 to pixel 4, the signal change from  pixel 2 to pixel 3, and so on with the influence getting smaller with the decay factor. In your code pixel 5 is influenced by the signal change from pixel 4 to pixel 5, the signal change from pixel 3 to pixel 5, the signal change from pixel 2 to pixel 5, and so on. I used only signal differences between neighboring pixels for estimating the model parameters, but your code calculates differences between non-neighboring pixels as well.

I think, something like that should produce the expected behavior:

```
difference = readoutRegister.head(lengthReadoutRegister - 1) - readoutRegister.tail(lengthReadoutRegister - 1);
            for (int deltaX = 0; deltaX <= frontEndElectronics->getOverAndUnderShootRange(); deltaX++)
            {
                totalContribution.head(lengthReadoutRegister - frontEndElectronics->getOverAndUnderShootRange()) += frontEndElectronics->getOverAndUnderShootStrength() * difference.subarray_shifted_by_deltaX * exp(-frontEndElectronics->getOverAndUnderShootDecayRate() * pow(deltaX, frontEndElectronics->getOverAndUnderShootDecaySpeed()));
            }
```

**Comment by SaraRegibo** [comment]
Hi @carstenpaproth, thanks for the feedback!  I'll look into it!

**Comment by SaraRegibo** [comment]
Apologies for the delay (I am in a meeting this week).  I just pushed the necessary changes.  They are available in Pull Request #404 (same as before).  Could you have a look, please?

**Comment by carstenpaproth** [comment]
@SaraRegibo 

Thanks again! Now, everything seems to work as expected. The over/undershoot effect is really small and there is almost no impact on the FGS.

**Comment by SaraRegibo** [comment]
Merged pull request #404.
